src/codegen/compile.ml

```diff
@@ -10352,97 +10766,116 @@ and compile_prim_invocation (env : E.t) ae p es at =
+      compile_exp_as env ae (SR.UnboxedWord32 Nat32) e ^^
       Prim.prim_word32toNat env
 
     | Int32, Int ->
       SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord32 e ^^
+      compile_exp_as env ae (SR.UnboxedWord32 Int32) e ^^
       Prim.prim_word32toInt env
 
     | Nat64, Nat ->
       SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord64 e ^^
+      compile_exp_as env ae (SR.UnboxedWord64 Nat64) e ^^
       BigNum.from_word64 env
 
     | Int64, Int ->
       SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord64 e ^^
+      compile_exp_as env ae (SR.UnboxedWord64 Int64) e ^^
       BigNum.from_signed_word64 env
 
     | Nat32, Char ->
-      SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord32 e ^^
-      TaggedSmallWord.check_and_tag_codepoint env
+      SR.UnboxedWord32 Type.Char, (* ! *)
+      compile_exp_as env ae (SR.UnboxedWord32 Nat32) e ^^
+      TaggedSmallWord.check_and_msb_adjust_codepoint env (* TBR *)
 
     | Float, Int ->
       SR.Vanilla,
       compile_exp_as env ae SR.UnboxedFloat64 e ^^
-      E.call_import env "rts" "bigint_of_float64"
+      E.call_import env "rts" "bigint_of_float64" ^^
+      let set_b, get_b = new_local env "b" in
+      (* adjust 31 to ubit scalar, done here to avoid touching bigint.rs *)
+      set_b ^^
+      get_b ^^
+      BitTagged.if_tagged_scalar env [I32Type]
+        (get_b ^^
+         compile_shrS_const 1l ^^
+         BigNum.from_signed_word32 env)
+        (get_b)
 
     | Int, Float ->
       SR.UnboxedFloat64,
       compile_exp_vanilla env ae e ^^
+      let set_b, get_b = new_local env "b" in
+      (* adjust ubit to 31 bit scalar, done here to avoid touching bigint.rs *)
+      set_b ^^
+      get_b ^^
+      BitTagged.if_tagged_scalar env [I32Type]
+        (get_b ^^
+         compile_shrS_const (Int32.of_int (32 - (BitTagged.ubits_of Type.Int))) ^^
+         compile_shl_const 1l)
+        (get_b) ^^
       E.call_import env "rts" "bigint_to_float64"
 
     | Float, Int64 ->
-      SR.UnboxedWord64,
+      SR.UnboxedWord64 Int64,
       compile_exp_as env ae SR.UnboxedFloat64 e ^^
       G.i (Convert (Wasm.Values.I64 I64Op.TruncSF64))
 
     | Int64, Float ->
       SR.UnboxedFloat64,
-      compile_exp_as env ae SR.UnboxedWord64 e ^^
+      compile_exp_as env ae (SR.UnboxedWord64 Int64) e ^^
       G.i (Convert (Wasm.Values.F64 F64Op.ConvertSI64))
     | Nat8, Nat16 ->
-      SR.Vanilla,
-      compile_exp_vanilla env ae e ^^
-      compile_shrU_const 8l
+      SR.UnboxedWord32 Nat16,
+      compile_exp_as env ae (SR.UnboxedWord32 Nat8) e ^^
+      TaggedSmallWord.lsb_adjust Nat8 ^^
+      TaggedSmallWord.msb_adjust Nat16
     | Nat16, Nat32 ->
-      SR.Vanilla,
-      compile_exp_vanilla env ae e ^^
-      compile_shrU_const 15l (* resulting Nat32 will always be unboxed *)
+      SR.UnboxedWord32 Nat32,
+      compile_exp_as env ae (SR.UnboxedWord32 Nat16) e ^^
+      TaggedSmallWord.lsb_adjust Type.Nat16
```

Comment:
👍

Reply:
Done, but there might be more such places...